File: frontend/lib/providers/test_results.dart
// Copyright 2023 Canonical Ltd.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License version 3, as
// published by the Free Software Foundation.
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//
// SPDX-FileCopyrightText: Copyright 2023 Canonical Ltd.
// SPDX-License-Identifier: GPL-3.0-only

import 'package:riverpod_annotation/riverpod_annotation.dart';

import '../models/test_result.dart';
import '../models/issue_attachment.dart';
import '../models/attachment_rule.dart';
import 'api.dart';

part 'test_results.g.dart';

@riverpod
class TestResults extends _$TestResults {
  @override
  Future<List<TestResult>> build(int testExecutionId) async {
    final api = ref.watch(apiProvider);
    return await api.getTestExecutionResults(testExecutionId);
  }

  Future<void> attachIssueToTestResult(
    int testResultId,
    int issueId, {
    AttachmentRule? attachmentRule,
  }) async {
    final api = ref.read(apiProvider);
    final attachedIssue = await api.attachIssue(
      issueId: issueId,
      testResultIds: [testResultId],
      attachmentRuleId: attachmentRule?.id,
    );
    final testResults = await future;
    final resultIndex =
        testResults.indexWhere((result) => result.id == testResultId);
    if (resultIndex == -1) {
      return;
    }
    final testResult = testResults[resultIndex];
    final attachments = testResult.issueAttachments;
    final alreadyAttached = attachments
        .any((attachment) => attachment.issue.id == attachedIssue.id);
    if (alreadyAttached) {
      return;
    }
    // Create a new attachment for the issue
    final newAttachment =
        IssueAttachment(issue: attachedIssue, attachmentRule: attachmentRule);
    final updatedAttachments = [...attachments, newAttachment];
    final updatedTestResult =
        testResult.copyWith(issueAttachments: updatedAttachments);
    final updatedTestResults = List<TestResult>.from(testResults);
    updatedTestResults[resultIndex] = updatedTestResult;
    state = AsyncData(updatedTestResults);
  }

  Future<void> detachIssueFromTestResult(int testResultId, int issueId) async {
    final api = ref.read(apiProvider);
    await api.detachIssue(
      issueId: issueId,
      testResultIds: [testResultId],
    );
    final testResults = await future;
    final resultIndex =
        testResults.indexWhere((result) => result.id == testResultId);
    if (resultIndex == -1) {
      return;
    }
    final testResult = testResults[resultIndex];
    final attachments = testResult.issueAttachments;
    final updatedAttachments = attachments
        .where((attachment) => attachment.issue.id != issueId)
        .toList();
    final updatedTestResult =
        testResult.copyWith(issueAttachments: updatedAttachments);
    final updatedTestResults = List<TestResult>.from(testResults);
    updatedTestResults[resultIndex] = updatedTestResult;
    state = AsyncData(updatedTestResults);
  }
}

File: backend/test_observer/controllers/reports/test_results.py
# Copyright 2024 Canonical Ltd.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License version 3, as
# published by the Free Software Foundation.
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# SPDX-FileCopyrightText: Copyright 2024 Canonical Ltd.
# SPDX-License-Identifier: AGPL-3.0-only

import csv
from datetime import datetime
from pathlib import Path

from fastapi import APIRouter, BackgroundTasks, Depends, Security
from fastapi.responses import FileResponse
from sqlalchemy import select
from sqlalchemy.orm import Session
from sqlalchemy.orm.attributes import InstrumentedAttribute

from test_observer.common.permissions import Permission, permission_checker
from test_observer.data_access.models import (
    Artefact,
    ArtefactBuild,
    Environment,
    TestCase,
    TestExecution,
    TestPlan,
    TestResult,
)
from test_observer.data_access.setup import get_db

router = APIRouter()

TESTRESULTS_REPORT_COLUMNS: list[InstrumentedAttribute] = [
    Artefact.family,
    Artefact.id,
    Artefact.name,
    Artefact.version,
    Artefact.status,
    Artefact.track,
    Artefact.stage,
    Artefact.series,
    Artefact.repo,
    Artefact.os,
    Artefact.series,
    Artefact.created_at,
    TestExecution.id,
    TestExecution.status,
    TestExecution.c3_link,
    TestExecution.checkbox_version,
    TestPlan.name,
    Environment.name,
    Environment.architecture,
    TestCase.template_id,
    TestCase.name,
    TestCase.category,
    TestResult.status,
    TestResult.created_at,
]


@router.get(
    "/test-results",
    response_class=FileResponse,
    dependencies=[Security(permission_checker, scopes=[Permission.view_report])],
)
def get_testresults_report(
    background_tasks: BackgroundTasks,
    start_date: datetime = datetime.min,
    end_date: datetime | None = None,
    db: Session = Depends(get_db),
):
    """
    Returns a csv report detailing all artefacts within a given date range. Together
    with their test executions and test results in csv format.
    """
    if end_date is None:
        end_date = datetime.now()

    cursor = db.execute(
        select(*TESTRESULTS_REPORT_COLUMNS)
        .join_from(Artefact, ArtefactBuild)
        .join_from(ArtefactBuild, TestExecution)
        .join_from(TestExecution, Environment)
        .join_from(TestExecution, TestPlan)
        .join_from(TestExecution, TestResult)
        .join_from(TestResult, TestCase)
        .where(Artefact.created_at >= start_date, Artefact.created_at <= end_date)
    )

    filename = "testresults_report.csv"

    try:
        with open(filename, "w") as csvfile:
            writer = csv.writer(csvfile)
            writer.writerow(TESTRESULTS_REPORT_COLUMNS)
            writer.writerows(cursor)
    finally:
        # Background tasks get called after the response was returned.
        # So effectively, this will delete the file after it was returned to the user
        background_tasks.add_task(lambda: Path(filename).unlink(missing_ok=True))

    return filename
